modules/raft/src/main/java/org/apache/ignite/internal/raft/RaftGroupServiceImpl.java

```diff
@@ -482,6 +482,13 @@ public ClusterService clusterService() {
         return cluster;
     }
 
+    @Override
+    public void updateConfiguration(PeersAndLearners configuration) {
```

Comment:
As @kgusakov mentioned given methods aren't thread safe. Several tickets will be created in order to fix this.

Reply:
Actually I meant that many methods from `RaftGroupServiceImpl` are not threadsafe, for example refreshMembers called in paralell with refreshMembers/addLearners/addPeers and other - can produce the client with the (peers, learners) pairs which is meaningless and weren't exist in any point of time in fact.

So, it is not connected with only this method, but I created the https://issues.apache.org/jira/browse/IGNITE-20738 and will mark this method with it.

Reply:
You can mark the whole class with this

Reply:
Marked the whole class, yep